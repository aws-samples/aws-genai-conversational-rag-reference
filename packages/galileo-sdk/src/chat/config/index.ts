/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */

import { BaseLanguageModel } from 'langchain/base_language';
import { PromptTemplate } from 'langchain/prompts';
import { difference } from 'lodash';
import { TKwags } from '../../common/types.js';
import { IModelInfo } from '../../models/types.js';
import { PromptRuntime } from '../../prompt/types.js';
import { ChainType } from '../../schema/index.js';
import { mergeConfig } from '../../utils/merge.js';
import { ChatEngineContext, ResolvedLLM } from '../context.js';
import { SearchRetrieverInput } from '../search.js';

export interface MemoryConfig {
  /**
   * Number of messages to fetch for contextual conversation history
   * - Set to 0 to disable conversation history (and condense question step), but still store messages
   */
  readonly limit?: number;
}

export interface LLMConfig {
  /**
   * LLM model, either predefined model uuid as string, or model info definition.
   */
  readonly model: string | IModelInfo | undefined;
  /** Additional model kwargs */
  readonly modelKwargs?: TKwags;
  /** Additional endpoint kwargs */
  readonly endpointKwargs?: TKwags;
}

export interface UnresolvedLLMChainConfig {
  readonly enabled?: boolean;
  readonly llm?: LLMConfig;
  readonly prompt?: string | PromptRuntime;
}

export interface LLMChainConfig {
  readonly enabled?: boolean;
  readonly llm: BaseLanguageModel;
  readonly prompt: PromptTemplate;
}

export interface UnresolvedChatEngineConfig {
  /** Indicates if config is the root, and should not be merged with ancestors */
  readonly root?: boolean;
  /** LLM to use for chains, unless overridden by the chain */
  readonly llm?: LLMConfig;
  /** Classify chain config, if undefined no classification will be performed */
  readonly classifyChain?: false | UnresolvedLLMChainConfig;
  /** Question/Answer chain config */
  readonly qaChain?: UnresolvedLLMChainConfig;
  /** Condense question chain (standalone question generator) */
  readonly condenseQuestionChain?: UnresolvedLLMChainConfig;

  readonly search: SearchRetrieverInput;
  readonly memory?: MemoryConfig;
}

export interface ChatEngineConfig {
  /** Indicates if config is the root, and should not be merged with ancestors */
  readonly root?: boolean;
  /** LLM to use for chains, unless overridden by the chain */
  readonly llm: BaseLanguageModel;
  /** Classify chain config, if undefined no classification will be performed */
  readonly classifyChain: false | LLMChainConfig;
  /** Question/Answer chain config */
  readonly qaChain: LLMChainConfig;
  /** Condense question chain (standalone question generator) */
  readonly condenseQuestionChain: LLMChainConfig;

  readonly search: SearchRetrieverInput;
  readonly memory?: MemoryConfig;
}

export function mergeUnresolvedChatEngineConfig(
  ...configs: Partial<UnresolvedChatEngineConfig>[]
): UnresolvedChatEngineConfig {
  return mergeConfig(...configs) as UnresolvedChatEngineConfig;
}

function extractPrivilegedKeys(obj: any, allowed: string[], prefix?: string): string[] {
  if (obj && typeof obj === 'object') {
    const privileged = difference(Object.keys(obj), allowed);
    if (privileged.length) {
      return prefix ? privileged.map((v) => `${prefix}.${v}`) : privileged;
    }
  }
  return [];
}

export const UNPRIVILEGED_KEYS = {
  MODEL: ['uuid'],
  SEARCH: ['filter', 'limit', 'scoreThreshold'],
};

export function extractPrivilegedChatEngineConfigKeys(config: Partial<UnresolvedChatEngineConfig>): string[] {
  const privileged: string[] = [];
  privileged.push(...extractPrivilegedKeys(config.llm?.model, UNPRIVILEGED_KEYS.MODEL, 'llm.model'));

  Object.values(ChainType).forEach((type) => {
    const key = getChainConfigKeyByType(type);
    const chainConfig = config[key] as UnresolvedLLMChainConfig | undefined;
    privileged.push(...extractPrivilegedKeys(chainConfig?.llm?.model, UNPRIVILEGED_KEYS.MODEL, `${key}.llm.model`));
  });

  privileged.push(...extractPrivilegedKeys(config.search, UNPRIVILEGED_KEYS.SEARCH, 'search'));

  return privileged;
}

export function assertNonPrivilegedChatEngineConfig(config: Partial<UnresolvedChatEngineConfig>): void {
  const privileged = extractPrivilegedChatEngineConfigKeys(config);

  if (privileged.length) {
    throw new Error(`ChatEngineConfig contains the following privileged properties: ${privileged.join(', ')}`);
  }
}

async function resolveLLM(
  llmConfig?: LLMConfig,
  defaultValues?: ResolvedLLM,
  options?: { verbose?: boolean },
): Promise<ResolvedLLM> {
  let modelInfo = defaultValues?.modelInfo;
  if (llmConfig?.model) {
    modelInfo = await ChatEngineContext.resolveModelInfo(llmConfig.model);
  }

  return ChatEngineContext.resolveLLM(modelInfo, { ...llmConfig, ...options });
}

export async function resolveChatEngineConfig(
  config: UnresolvedChatEngineConfig,
  options?: { verbose?: boolean },
): Promise<ChatEngineConfig> {
  const defaultLLM = await resolveLLM(config.llm, undefined, options);
  const qaLLM = await resolveLLM(config.qaChain?.llm, defaultLLM, options);
  const condensedLLM = await resolveLLM(config.condenseQuestionChain?.llm, defaultLLM, options);
  const classifyLLM =
    config.classifyChain && config.classifyChain.enabled === true
      ? await resolveLLM(
          {
            model: defaultLLM.modelInfo,
            modelKwargs: { temperature: 0 },
            ...config.classifyChain?.llm,
          },
          defaultLLM,
          options,
        )
      : false;
  const classify = classifyLLM && config.classifyChain;

  return {
    ...config,
    llm: defaultLLM.llm,
    qaChain: {
      // required
      enabled: true,
      llm: qaLLM.llm,
      prompt: await ChatEngineContext.resolvePromptTemplate(
        ChainType.QA,
        qaLLM.adapter.prompt?.chat?.QA,
        config.qaChain?.prompt,
      ),
    },
    condenseQuestionChain: {
      // enabled by default
      enabled: config.condenseQuestionChain?.enabled ?? true,
      llm: condensedLLM.llm,
      prompt: await ChatEngineContext.resolvePromptTemplate(
        ChainType.CONDENSE_QUESTION,
        condensedLLM.adapter.prompt?.chat?.CONDENSE_QUESTION,
        config.condenseQuestionChain?.prompt,
      ),
    },
    classifyChain: classify
      ? {
          // disabled by default
          enabled: config.classifyChain.enabled ?? false,
          llm: classifyLLM.llm,
          prompt: await ChatEngineContext.resolvePromptTemplate(
            ChainType.CLASSIFY,
            classifyLLM.adapter.prompt?.chat?.CLASSIFY,
            config.classifyChain?.prompt,
          ),
        }
      : false,
  };
}

export function getChainConfigKeyByType(type: ChainType) {
  switch (type) {
    case ChainType.QA:
      return 'qaChain';
    case ChainType.CONDENSE_QUESTION:
      return 'condenseQuestionChain';
    case ChainType.CLASSIFY:
      return 'classifyChain';
  }
}

export function getChainConfigByType<T extends UnresolvedChatEngineConfig | ChatEngineConfig>(
  type: ChainType,
  config?: T,
) {
  return config && config[getChainConfigKeyByType(type)];
}
