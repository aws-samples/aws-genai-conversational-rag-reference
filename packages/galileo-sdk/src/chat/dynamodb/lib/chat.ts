/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
import {
  DynamoDBDocumentClient,
  PutCommand,
  QueryCommandInput,
  DeleteCommand,
  UpdateCommand,
} from '@aws-sdk/lib-dynamodb';
import { v4 as uuidv4 } from 'uuid';
import { getAllChatMessageIds } from './messages.js';
import { listChatMessageSources } from './sources.js';
import {
  getChatKey,
  getChatsByTimeKey,
  DDBChat,
  bulkDelete,
  Keys,
  DDBUpdateOutput,
  getAllByPagination,
} from './util.js';

export async function createChat(
  documentClient: DynamoDBDocumentClient,
  tableName: string,
  userId: string,
  title: string,
) {
  const newChatSessionId = uuidv4();

  const timestamp = Date.now();
  const keys = getChatKey(userId, newChatSessionId);
  const gsiKeys = getChatsByTimeKey(userId, `${timestamp}`);
  const chat: DDBChat = {
    chatId: newChatSessionId,
    title: title,
    createdAt: timestamp,
    userId: userId,
    ...keys,
    ...gsiKeys,
    entity: 'CHAT',
  };

  const command = new PutCommand({
    TableName: tableName,
    Item: chat,
    ReturnValues: 'NONE',
  });

  const response = await documentClient.send(command);

  return {
    response,
    chat,
  };
}

export async function listAllChatsByTime(
  documentClient: DynamoDBDocumentClient,
  tableName: string,
  indexName: string,
  userId: string,
  asc = false,
) {
  const keys = getChatsByTimeKey(userId);

  const commandInput: QueryCommandInput = {
    TableName: tableName,
    IndexName: indexName,
    KeyConditionExpression: 'GSI1PK = :PK',
    ExpressionAttributeValues: {
      ':PK': keys.GSI1PK,
    },
    ScanIndexForward: asc,
  };

  // We've decided to return all chats for a user and then filter and paginate client side
  return getAllByPagination<DDBChat>(documentClient, commandInput);
}

export async function updateChat(
  documentClient: DynamoDBDocumentClient,
  tableName: string,
  userId: string,
  chatId: string,
  title: string,
) {
  const keys = getChatKey(userId, chatId);

  const command = new UpdateCommand({
    TableName: tableName,
    Key: keys,
    ConditionExpression: 'attribute_exists(PK) and attribute_exists(SK)',
    UpdateExpression: 'set title = :title',
    ExpressionAttributeValues: {
      ':title': title,
    },
    ReturnValues: 'ALL_NEW',
  });

  const result = (await documentClient.send(command)) as DDBUpdateOutput<DDBChat>;
  return result.Attributes;
}

async function deleteChat(documentClient: DynamoDBDocumentClient, tableName: string, userId: string, chatId: string) {
  const keys = getChatKey(userId, chatId);

  const command = new DeleteCommand({
    TableName: tableName,
    Key: keys,
    ReturnValues: 'ALL_OLD',
  });

  return documentClient.send(command);
}

export async function deleteChatAndMessages(
  documentClient: DynamoDBDocumentClient,
  tableName: string,
  indexName: string,
  userId: string,
  chatId: string,
) {
  // For now we'll perform all the delete actions in a single lambda
  // We might want to consider creating a job queue for deleting the chatMessages async
  const chatMessageKeys = await getAllChatMessageIds(documentClient, tableName, indexName, userId, chatId);

  let chatMessageSourceKeys: Keys[] = [];
  for (const messageKeys of chatMessageKeys) {
    const response = await listChatMessageSources(documentClient, tableName, userId, messageKeys.messageId);
    if (response) {
      chatMessageSourceKeys = [...chatMessageSourceKeys, ...response];
    }
  }

  const deleteKeys = [...chatMessageKeys, ...chatMessageSourceKeys].map((k) => ({ PK: k.PK, SK: k.SK }));

  console.log(`found ${chatMessageKeys.length} messages`);
  console.log(`total records to delete: ${deleteKeys.length}`);

  if (chatMessageKeys.length > 0) {
    await bulkDelete(documentClient, tableName, deleteKeys);
  }
  await deleteChat(documentClient, tableName, userId, chatId);
}
