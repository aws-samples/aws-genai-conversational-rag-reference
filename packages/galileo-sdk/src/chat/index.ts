/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
import '../langchain/patch.js';
export * as dynamodb from './dynamodb/index.js';
export * as chain from './chain.js';
export * as context from './context.js';
export * as memory from './memory.js';

export * from './engine.js';
