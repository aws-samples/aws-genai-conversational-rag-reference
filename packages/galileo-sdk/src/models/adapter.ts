/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
import { BaseSageMakerContentHandler } from 'langchain/llms/sagemaker_endpoint';
import { set, get, isEmpty } from 'lodash';
import { BaseChatTemplatePartials } from '../prompt/templates/chat/base.js';
import { ChatTemplateTypedRuntimeRecord, PromptTemplateStore } from '../prompt/templates/store/registry.js';
import { ChainType } from '../schema/index.js';

/**
 * @struct
 */
export interface IContentHandlerAdapterInput {
  /**
   * Path in input object to define the prompt value.
   * @see https://lodash.com/docs/#set
   * @default "text_inputs"
   */
  readonly promptKey?: string;
  /**
   * Path in the input object to define the model kwargs.
   * @see https://lodash.com/docs/#set
   * @default undefined - Kwargs will be spread on root
   */
  readonly modelKwargsKey?: string;
  // readonly prepend?: string;
  // readonly append?: string;
  /**
   * @default "utf-8"
   */
  readonly encoding?: BufferEncoding;
}
/**
 * @struct
 */
export interface IContentHandlerAdapterOutput {
  /**
   * Path of response object to extract as output.
   * @see https://lodash.com/docs/#get
   * @default `[0].generated_text`
   */
  readonly jsonpath?: string;
  /**
   * @default "utf-8"
   */
  readonly encoding?: BufferEncoding;
}
/**
 * @struct
 */
export interface IContentHandlerAdapter {
  readonly contentType?: string;
  readonly accepts?: string;
  readonly input?: IContentHandlerAdapterInput;
  readonly output?: IContentHandlerAdapterOutput;
}

/**
 * @struct
 */
export type IChatPromptAdapter = {
  readonly base?: Partial<BaseChatTemplatePartials>;
} & Partial<ChatTemplateTypedRuntimeRecord>;
/**
 * @struct
 */
export interface IPromptAdapter {
  readonly chat?: IChatPromptAdapter;
}

export class PromptAdapter implements IPromptAdapter {
  private static resolveRuntime(
    type: ChainType,
    base?: Partial<BaseChatTemplatePartials>,
    runtime?: string | any,
  ): any | undefined {
    if (runtime == null) {
      if (base == null) return undefined;
      return {
        templatePartials: {
          ...base,
        },
      };
    }

    if (typeof runtime === 'string') {
      // TODO: support non-system templates once we implement store
      // this will throw error if invalid template scope/type/substype
      const parsedId = PromptTemplateStore.parseId(runtime, {
        scope: PromptTemplateStore.SYSTEM_SCOPE,
        type: PromptTemplateStore.CHAT_TYPE,
        subtype: type,
      });
      runtime = PromptTemplateStore.getSystemChatTemplateRuntime(type, parsedId.name);
    }

    return {
      ...runtime,
      templatePartials: {
        ...base,
        ...runtime.templatePartials,
      },
    };
  }

  readonly chat?: IChatPromptAdapter;

  constructor(adapter?: IPromptAdapter) {
    const { base, ...typed } = adapter?.chat || {};

    this.chat = {
      base,
      ...Object.fromEntries(
        Object.values(ChainType).map((type) => {
          if (type in typed) {
            return [type, PromptAdapter.resolveRuntime(type, base, typed[type])];
          } else if (!isEmpty(base)) {
            return [type, PromptAdapter.resolveRuntime(type, base)];
          } else {
            return [type, undefined];
          }
        }),
      ),
    };
  }
}

/**
 * @struct
 */
export interface IBaseModelAdapter {
  readonly contentHandler?: IContentHandlerAdapter;
}

/**
 * Model adapter definition for inference
 * @struct
 */
export interface IModelAdapter extends IBaseModelAdapter {
  readonly prompt?: IPromptAdapter;
}

/**
 * Model adapter for embedding
 * @struct
 */
export interface IEmbeddingModelAdapter extends IBaseModelAdapter {}

export const DEFAULT_OUTPUT_JSONPATH = '[0].generated_text';

export class AdaptedContentHandler extends BaseSageMakerContentHandler<string, string> {
  readonly contentType: string;
  readonly accepts: string;

  protected readonly promptKey: string;
  protected readonly modelKwargsKey?: string;
  protected readonly inputEncoding: BufferEncoding;

  protected readonly outputJsonpath: string;
  protected readonly outputEncoding: BufferEncoding;

  constructor(adapter?: IContentHandlerAdapter) {
    super();

    this.accepts = adapter?.accepts ?? 'application/json';
    this.contentType = adapter?.contentType ?? 'application/json';

    this.promptKey = adapter?.input?.promptKey ?? 'text_inputs';
    this.modelKwargsKey = adapter?.input?.modelKwargsKey;
    this.inputEncoding = adapter?.input?.encoding ?? 'utf-8';

    this.outputJsonpath = adapter?.output?.jsonpath ?? DEFAULT_OUTPUT_JSONPATH;
    this.outputEncoding = adapter?.output?.encoding ?? 'utf-8';

    if (this.accepts !== 'application/json' || this.contentType != 'application/json') {
      throw new Error(
        `Only "application/json" is supported at this time: ${this.accepts} / ${this.contentType} not implemented`,
      );
    }
  }

  async transformInput(prompt: string, modelKwargs: Record<string, unknown>): Promise<Uint8Array> {
    const input = {};

    set(input, this.promptKey, prompt);

    if (this.modelKwargsKey) {
      set(input, this.modelKwargsKey, modelKwargs);
    } else {
      // spread kwargs if not key
      Object.assign(input, modelKwargs);
    }

    console.debug('AdaptedContentHandler#transformInput:', { prompt, modelKwargs, input });

    return Buffer.from(JSON.stringify(input), this.inputEncoding);
  }

  async transformOutput(output: Uint8Array): Promise<string> {
    const response = JSON.parse(Buffer.from(output).toString(this.outputEncoding));
    // extract the path from json
    const result = get(response, this.outputJsonpath);
    console.debug('AdaptedContentHandler#transformOutput:', { response, result });
    return result;
  }
}

export class ModelAdapter {
  readonly prompt?: IPromptAdapter;

  readonly contentHandler: AdaptedContentHandler;

  readonly isDefault: boolean;

  constructor(config?: IModelAdapter) {
    this.isDefault = config == null || isEmpty(config);

    this.prompt = new PromptAdapter(config?.prompt);

    this.contentHandler = new AdaptedContentHandler(config?.contentHandler);
  }
}
