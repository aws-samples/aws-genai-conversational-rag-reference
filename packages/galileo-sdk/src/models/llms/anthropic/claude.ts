/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
import type { IModelAdapter } from '../../adapter.js';
import { DEFAULT_MAX_NEW_TOKENS } from '../../constants.js';
import type { Kwargs } from '../../types.js';

// https://docs.aws.amazon.com/bedrock/latest/userguide/prompt-templates-and-examples.html#qa-with-context
export const CLAUDE_V2_ADAPTER: IModelAdapter = {
  prompt: {
    chat: {
      questionAnswer: {
        template: `\n\nHuman: Read the following chat history inside <history></history> XML tags, and then rephrase the "Followup Question" below to be a concise standalone question in its original language. Without answering the question, return only the standalone question:

<history>
{{>ChatHistory}}
</history>
Followup Question: {{question}}

Assistant: `,
      },
      condenseQuestion: {
        template: `\n\nHuman: Read the follow text inside <text></text> XML tags, and then answer the question based on the provided rules inside <rules></rules>:

<text>
{{corpus}}
</text>
<rules>
1. only use knowledge from the provided corpus to answer the question
2. always be truthful, honest, unbiased, and unharmful
3. be concise, do not repeat the question or yourself in the answer
</rules>
Based on the text above, {{question}}

Assistant: `,
      },
    },
  },
  contentHandler: {
    input: {
      promptKey: 'prompt',
    },
    output: {
      jsonpath: 'completion',
    },
  },
};

export const CLAUDE_V2_KWARGS: Kwargs = {
  temperature: 0,
  max_tokens_to_sample: DEFAULT_MAX_NEW_TOKENS,
  stop_sequences: ['\n\nHuman:'],
};
