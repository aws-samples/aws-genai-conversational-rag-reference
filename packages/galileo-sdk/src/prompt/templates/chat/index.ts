/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
import CLASSIFY from './classify.js';
import CONDENSE_QUESTION from './condense-question.js';
import QA from './question-answer.js';
import { ChainType } from '../../../schema/index.js';
import { PromptRuntime } from '../../types.js';

export const CHAT_PROMPT_TEMPLATES: Record<ChainType, PromptRuntime> = {
  CLASSIFY,
  CONDENSE_QUESTION,
  QA,
};

export default CHAT_PROMPT_TEMPLATES;

export { QA, CLASSIFY, CONDENSE_QUESTION };
