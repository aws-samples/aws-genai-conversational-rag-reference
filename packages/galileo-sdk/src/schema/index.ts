/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
export enum ChainType {
  QA = 'QA',
  CONDENSE_QUESTION = 'CONDENSE_QUESTION',
  CLASSIFY = 'CLASSIFY',
}
