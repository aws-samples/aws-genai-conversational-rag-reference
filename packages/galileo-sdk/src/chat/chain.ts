/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
import { BaseLanguageModel } from 'langchain/base_language';
import { CallbackManagerForChainRun } from 'langchain/callbacks';
import { BaseChain, ChainInputs, LLMChain, loadQAChain, QAChainParams, StuffDocumentsChain } from 'langchain/chains';
import { PromptTemplate } from 'langchain/prompts';
// import { SerializedChatVectorDBQAChain } from "./serde.js";
import { ChainValues } from 'langchain/schema';
import { BaseRetriever } from 'langchain/schema/retriever';
import { getLogger } from '../common/index.js';
import { startPerfMetric } from '../common/metrics/index.js';
import { PojoOutputParser } from '../langchain/output_parsers/pojo.js';

const logger = getLogger('chat/chain');

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type LoadValues = Record<string, any>;

export interface ChatEngineChainInput extends ChainInputs {
  retriever: BaseRetriever;
  /**
   * Classify chain for adaptive chain (classification, category, language, etc.)
   * which outputs JSON config proved to following chain inputs
   */
  classifyChain: false | LLMChain;
  /**
   * Primary Question/Answer chain that provides the final answer generation to the user.
   */
  qaChain: BaseChain;
  /**
   * Followup question generator chain which condenses followup question based on chat history
   * into a standalone question which is provided to the QA chain.
   */
  condenseQuestionChain: LLMChain;
  returnSourceDocuments?: boolean;
  inputKey?: string;
}

export type ChatEngineChainFromInput = {
  returnSourceDocuments?: boolean;
  retriever: BaseRetriever;
  classifyChain:
    | false
    | {
        llm: BaseLanguageModel;
        prompt: PromptTemplate;
        /** Key to use for output, default to "classification" */
        outputKey?: string;
      };
  condenseQuestionChain: {
    llm: BaseLanguageModel;
    prompt: PromptTemplate;
  };
  qaChain: {
    // TODO: support other combine document types, we need to implement templates + adapters?
    type: 'stuff';
    llm: BaseLanguageModel;
    prompt: PromptTemplate;
  };
} & Omit<
  ChatEngineChainInput,
  'retriever' | 'combineDocumentsChain' | 'qaChain' | 'classifyChain' | 'condenseQuestionChain'
>;

export class ChatEngineChain extends BaseChain implements ChatEngineChainInput {
  static lc_name() {
    return 'ChatEngineChain';
  }

  /**
   * Static method to create a new ChatEngineChain from a
   * BaseLanguageModel and a BaseRetriever.
   * @param retriever {@link BaseRetriever} instance used to retrieve relevant documents.
   * @param options.returnSourceDocuments Whether to return source documents in the final output
   * @param options.condenseQuestionChain Options to initialize the standalone question generation chain used as the first internal step
   * @param options.qaChain {@link QAChainParams} used to initialize the QA chain used as the second internal step
   * @returns A new instance of ChatEngineChain.
   */
  static from(options: ChatEngineChainFromInput): ChatEngineChain {
    const {
      qaChain: qaChainOptions,
      classifyChain: classifyOptions,
      condenseQuestionChain: condenseQuestionChainOptions,
      verbose,
      ...rest
    } = options;

    const qaChain = loadQAChain(qaChainOptions.llm, {
      verbose,
      ...qaChainOptions,
    });

    const classifyChain =
      classifyOptions &&
      new LLMChain({
        verbose,
        ...classifyOptions,
        outputKey: 'classification',
        // classify chain must return parsable JSON
        outputParser: new PojoOutputParser<any>(),
      });

    const condenseQuestionChain = new LLMChain({
      verbose,
      ...condenseQuestionChainOptions,
    });

    const instance = new this({
      classifyChain,
      qaChain,
      condenseQuestionChain,
      verbose,
      ...rest,
    });
    return instance;
  }

  inputKey = 'question';

  classificationKey = 'classification';

  chatHistoryKey = 'chat_history';

  get inputKeys() {
    return [this.inputKey, this.chatHistoryKey];
  }

  get outputKeys() {
    return this.qaChain.outputKeys.concat(this.returnSourceDocuments ? ['sourceDocuments'] : []);
  }

  get traceData(): any {
    return this._traceData;
  }

  retriever: BaseRetriever;

  classifyChain: false | LLMChain;

  qaChain: BaseChain;

  condenseQuestionChain: LLMChain;

  returnSourceDocuments = false;

  protected _traceData?: any;

  constructor(fields: ChatEngineChainInput) {
    super(fields);
    this.retriever = fields.retriever;
    this.classifyChain = fields.classifyChain;
    this.qaChain = fields.qaChain;
    this.condenseQuestionChain = fields.condenseQuestionChain;
    this.inputKey = fields.inputKey ?? this.inputKey;
    this.returnSourceDocuments = fields.returnSourceDocuments ?? this.returnSourceDocuments;
  }

  /** @ignore */
  async _call(values: ChainValues, runManager?: CallbackManagerForChainRun): Promise<ChainValues> {
    if (!(this.inputKey in values)) {
      throw new Error(`Question key ${this.inputKey} not found.`);
    }
    if (!(this.chatHistoryKey in values)) {
      throw new Error(`Chat history key ${this.chatHistoryKey} not found.`);
    }
    const question: string = values[this.inputKey];
    const chatHistory = values[this.chatHistoryKey] || [];

    let classification: ChainValues | undefined;
    if (this.classifyChain) {
      logger.debug('Calling classify chain: ', { question });
      const $$ClassifyChainExecutionTime = startPerfMetric('ClassifyChainExecutionTime');
      classification = (await this.classifyChain.call({ question }))[this.classificationKey];
      $$ClassifyChainExecutionTime();
      logger.debug('Result from classify chain: ', { classification });
    }

    let newQuestion = classification?.question || question;
    const hasHistory = chatHistory.length > 0;
    if (hasHistory) {
      const condenseQuestionInput: ChainValues = {
        question: newQuestion,
        ...classification,
        chat_history: chatHistory,
      };
      logger.debug('Chain:condenseQuestionChain:input', { input: condenseQuestionInput });
      const $$QuestionGeneratorExecutionTime = startPerfMetric('QuestionGeneratorExecutionTime');
      const result = await this.condenseQuestionChain.call(
        condenseQuestionInput,
        runManager?.getChild('question_generator'),
      );
      $$QuestionGeneratorExecutionTime();
      logger.debug('Chain:condenseQuestionChain:output', { output: result });

      const keys = Object.keys(result);
      if (keys.length === 1) {
        newQuestion = result[keys[0]];
        logger.debug(`Rewrote question from "${question}" to "${newQuestion}`);
      } else {
        throw new Error('Return from llm chain has multiple values, only single values supported.');
      }
    }
    logger.debug('Chain:retriever:getRelevantDocuments:query', { query: newQuestion });
    const $$GetRelevantDocumentsExecutionTime = startPerfMetric('GetRelevantDocumentsExecutionTime');
    const docs = await this.retriever.getRelevantDocuments(newQuestion, runManager?.getChild('retriever'));
    $$GetRelevantDocumentsExecutionTime();

    const inputs = {
      ...classification,
      input_documents: docs,
      chat_history: chatHistory,
      question: newQuestion,
    };

    logger.debug('Chain:condenseQuestionChain:input', { input: inputs });
    const $$CombineDocumentsExecutionTime = startPerfMetric('CombineDocumentsExecutionTime');
    const result = await this.qaChain.call(inputs, runManager?.getChild('combine_documents'));
    $$CombineDocumentsExecutionTime();
    logger.debug('Chain:condenseQuestionChain:output', { output: result });

    this._traceData = {
      originalQuestion: question,
      standaloneQuestion: newQuestion,
      classification,
      hasHistory,
      chainValues: values,
      chatHistory,
      sourceDocuments: docs,
      inputs,
      result,
      chains: {
        qaChain: this.qaChain instanceof StuffDocumentsChain ? this.qaChain.llmChain.toJSON() : this.qaChain.toJSON(),
        condenseQuestionChain: this.condenseQuestionChain.toJSON(),
        classifyChain: this.classifyChain && this.classifyChain.toJSON(),
      },
    };

    logger.debug('Trace data', { traceData: this.traceData });

    if (this.returnSourceDocuments) {
      return {
        ...result,
        sourceDocuments: docs,
      };
    }
    return result;
  }

  _chainType(): string {
    return 'conversational_retrieval_chain';
  }

  // static async deserialize(
  //   _data: SerializedChatVectorDBQAChain,
  //   _values: LoadValues
  // ): Promise<ChatEngineChain> {
  //   throw new Error("Not implemented.");
  // }

  // serialize(): SerializedChatVectorDBQAChain {
  //   throw new Error("Not implemented.");
  // }
}
