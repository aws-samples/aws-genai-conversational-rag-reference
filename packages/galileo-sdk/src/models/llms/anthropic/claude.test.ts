/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
// @ts-ignore - .test files are ignored
import type {} from '@types/jest';
import { AIMessage, HumanMessage, SystemMessage } from 'langchain/schema';
import { CLAUDE_V2_ADAPTER } from './claude';
import { ChatCondenseQuestionPromptTemplate, ChatQuestionAnswerPromptTemplate } from '../../../prompt/templates/chat';
import { ModelAdapter } from '../../adapter';

describe('models/llms/anthropic/claude', () => {
  describe('adapter', () => {
    const adapter = new ModelAdapter(CLAUDE_V2_ADAPTER);

    test('should render qa prompt', async () => {
      const template = new ChatQuestionAnswerPromptTemplate({
        ...adapter.prompt?.chat?.questionAnswer,
      });
      expect(
        await template.format({
          question: 'THE QUESTION',
          domain: 'DOMAIN',
          context: 'THE CORPUS',
        }),
      ).toMatchSnapshot();
    });

    test('should render condense prompt', async () => {
      const template = new ChatCondenseQuestionPromptTemplate({
        ...adapter.prompt?.chat?.condenseQuestion,
      });
      expect(
        await template.format({
          question: 'THE QUESTION',
          chat_history: [new SystemMessage('A system message'), new HumanMessage('Hi'), new AIMessage('Hello!')],
        }),
      ).toMatchSnapshot();
    });
  });
});
