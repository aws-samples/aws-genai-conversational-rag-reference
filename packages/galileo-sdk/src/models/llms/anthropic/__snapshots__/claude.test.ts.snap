// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`models/llms/anthropic/claude adapter should render condense prompt 1`] = `
"

Human: Read the following chat history inside <history></history> XML tags, and then rephrase the "Followup Question" below to be a concise standalone question in its original language. Without answering the question, return only the standalone question:

<history>
System: A system message

Human: Hi

Assistant: Hello!

</history>
Followup Question: THE QUESTION

Assistant: "
`;

exports[`models/llms/anthropic/claude adapter should render qa prompt 1`] = `
"

Human: Read the follow text inside <text></text> XML tags, and then answer the question based on the provided rules inside <rules></rules>:

<text>
THE CORPUS
</text>
<rules>
1. only use knowledge from the provided corpus to answer the question
2. always be truthful, honest, unbiased, and unharmful
3. be concise, do not repeat the question or yourself in the answer
</rules>
Based on the text above, THE QUESTION

Assistant: "
`;
